File: Src/ButterflyNet.Parser/Strategies/HorizontalRulerStrategy.cs
using System.Text.RegularExpressions;
using ButterflyNet.Parser.Satisfiers;
using ButterflyNet.Parser.Scopes;

namespace ButterflyNet.Parser.Strategies {
	public class HorizontalRulerStrategy : ScopeDrivenStrategy {
		public HorizontalRulerStrategy() {
			AddSatisfier<StartOfLineSatisfier>();
			AddSatisfier<CannotNestInsideInlineSatisfier>();
			AddSatisfier<HRuleSatisfier>();
		}

		//needs to be less than strike through, which is three dashes in a row
		public override int Priority { get { return DefaultPriority - 2; } }

		protected override void DoExecute(ParseContext context) {
			context.Input.Read(3);
			context.UpdateCurrentChar();

			OpenScope(new HorizontalRulerScope(), context);
			CloseCurrentScope(context);
		}

		private class HRuleSatisfier : ISatisfier {
			private static readonly Regex regex = new Regex(@"^----(?:\n|$)");
			public bool IsSatisfiedBy(ParseContext context) {
				return regex.IsMatch(context.Input.Substring);
			}
		}
	}
}

File: src/js/src/strategies/HorizontalRulerStrategy.js
function HorizontalRulerStrategy() {
	HorizontalRulerStrategy.$parent.call(this);
	
	this.addSatisfier(new StartOfLineSatisfier());
	this.addSatisfier(new CannotNestInsideInlineSatisfier());
	this.addSatisfier(function(context) { return /^----(?:\n|$)/.test(context.input.substring()); });
	
	this.priority = ParseStrategy.defaultPriority - 2;
	
	this.doExecute = function(context) {
		context.input.read(3);
		context.updateCurrentChar();
		
		this.openScope(new HorizontalRulerScope(), context);
		this.closeCurrentScope(context);
	};
}

extend(ScopeDrivenStrategy, HorizontalRulerStrategy);